### PR TITLE
WIP: Use sd-bus to connect to the system bus

### PR DESCRIPTION
**Pull request by stefwalter**
This is an example pull request of how to use sd-bus and gdbus in the same storaged process.

This allows us to remove an obstacle to contribution. GObject and GDBus are awkward ... sd-bus is much more straight forward.

Remaining work:
- [ ] Comments and documentation
- [ ] Make a combined Object Manager
- [ ] Figure out the configure.ac changes better.
- [x] Squash and rebase.

### CODE REVIEW

File: src/storageddbus.c [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+  g_assert (system != NULL);
+
+  r = sd_bus_default_system (system);
+  if (r < 0)
+    {
+      g_critical ("Couldn't open system DBus bus: %s", g_strerror (-r));
+      return FALSE;
+    }
+
+  if (sd_bus_attach_event (*system, NULL, SD_EVENT_PRIORITY_NORMAL) < 0)
```
Comment (vpodzime) [inline src/storageddbus.c]: I cannot find any piece of documentation for this function, but shouldn't this pass some event to it?

File: src/storageddbus.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+#ifndef __STORAGED_DBUS_H__
+#define __STORAGED_DBUS_H__
+
+#include <systemd/sd-bus.h>
+#include <gio/gio.h>
+
+gboolean                storaged_dbus_initialize                 (sd_bus **system,
+                                                                  GDBusConnection **connection);
```
Comment (vpodzime) [inline src/storageddbus.h]: any reason for this blank space?

File: src/storageddbus.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+/*
+ * These GDBus related functions are unsupported due to the sd-bus to GDBus
+ * bridge. In particular the GDBus code will believe that it has a different
+ * unique name than actually present on the bus.
+ */
+
+#define g_dbus_connection_get_unique_name unsupported__dbus_connection_get_unique_name
```
Comment (vpodzime) [inline src/storageddbus.h]: Is ``unsupported__dbus_connection_get_unique_name`` defined somewhere? Or is this just to cause things crash?

**Comment by stefwalter** [comment]
@dwlehman and I discussed various obstacles to storage folks contributing to storaged. One of those obstacles was GObject ugliness and GDBusSkeleton nastiness. This results in a lot of boiler plate. I said I would try to get sd-bus  working, which ends up being much simpler to add API, expose it via an ObjectManager.

Rather than a "rewrite everything all at once" approach (which is a fail) ... a more measured approach such as this allows refactoring bit by bit.

One of the requirements that @dwlehman was suggesting was that any refactored section have full tests before refactoring takes place ... that's an aside ... and not necessarily related to this.

So in short, the goal of this pull request, once finished, is to allow better contribution by storage experts to storaged ... and remove a big allergy towards contribution.

**Comment by vpodzime** [comment]
Thanks for the details @stefwalter! Sounds very reasonable to me. However, I must say that these changes aren't really convincing me that sd-bus is any (neither a lot) better. ;) I'm looking forward to seeing more here.

**Comment by StorageGhoul** [comment]
Can one of the admins verify this patch?

**Comment by StorageGhoul** [comment]
Can one of the admins verify this patch?

**Comment by vpodzime** [comment]
I don't feel like this is going to get attention any time soon. I'm closing this PR, feel free to reopen or resubmit it once there's some traction again.
